subsys/net/ip/packet_socket.c

```diff
@@ -35,8 +35,22 @@ enum net_verdict net_packet_socket_input(struct net_pkt *pkt)
 	 * socket combination packet has to be feed to raw socket and only
 	 * data part to be feed to non raw socket.
 	 */
+	enum net_verdict ret;
 
-	net_pkt_set_family(pkt, AF_PACKET);
+	if (net_pkt_family(pkt) == AF_PACKET) {
```

Comment:
I just wonder where is the family set to be `AF_PACKET`. This function is called at the very beginning of the input processing, at this point we do not know the address family yet.

Reply:
I wanted to optimize the code to avoid some unnecessary cloning of packets. However, as AF_PACKET cannot be set so early the packet is cloned always anyway.  So it seems like I shall just remove the if from line 40 above.

Reply:
I think we could clone the packet only if `net_conn_input()` finds out that the packet needs to be fed to UDP/TCP socket. No need to do it always if there is no socket listening those packets.

Reply:
So perhaps we could feed the packet to `net_conn_input()` and if it notices that there is other than `AF_PACKET` listening data, it would return the original net_pkt (so that `NET_CONTINUE` is returned by `net_packet_socket_input()`), and let the `net_conn_input()` clone the packet and pass the cloned one to raw socket.

Reply:
I've updated patches according to your above guidelines.